common/flatpak-run.c

```diff
@@ -160,6 +177,18 @@ write_xauth (char *number, FILE *output)
               local_xa.number_length = 2;
             }
 
+          if (local_xa.family == FamilyLocal &&
+              !auth_streq (unames.nodename, local_xa.address, local_xa.address_length))
+            {
+              /* If we decided to propagate this cookie, but its address
+               * doesn't match `unames.nodename`, we need to change it or
+               * inside the container it will not work.
+               */
+              g_free (local_xa.address);
+              local_xa.address = g_strdup (unames.nodename);
```

Comment:
This seems suspicious to me: it seems like it's relying on undefined behaviour, at best. We have something like this:

* xa = XauReadAuth (f)
    * libXau allocates an Xauth struct
    * libXau allocates strings `address`, `number`, `name` and/or `data` in it
* local_xa = *xa
    * local_xa is on the stack
    * local_xa.address, local_xa.number, local_xa.name, local_xa.data are "borrowed" from xa
* free local_xa.address
    * now local_xa.address and xa->address are both dangling pointers to freed memory (!)
* copy unames.nodename into local_xa.address
    * now local_xa.address is valid again
    * xa->address is still a dangling pointer to freed memory
* local_xa goes out of scope
    * local_xa.address is leaked
* XauDisposeAuth (xa)
    * The dangling pointer xa->address is freed again - a double-free

I would have expected something more like:

```
              local_xa.address = unames.nodename;
              local_xa.address_length = strlen (local_xa.address);
```

or maybe, if we do need to copy the address,

```
              g_autofree char *nodename = NULL;
...
              nodename = g_strdup (unames.nodename);
              local_xa.address = nodename;
              local_xa.address_length = strlen (local_xa.address);
```

Reply:
Ah you are right.
I tested it on OpenSUSE but I did multiple iterations, so it might very well be that the version I tested was before I introduced the free for the old `local_xa.address`.

I'll test it again and I'll push a new revision.

Reply:
I confirmed the invalid `free()` with valgrind. It should be fixed in the latest revision.